Issue 65 replace all font styling with global font classes

SeiKasahara opened the pull request:
## Change Summary

1. Reworked Signup page.
2. Created new components and tested the dummy data to ensure we can register.
3. Created boundary check part.
4. Created regex pattern to check user's input

## Related Issue
#65 #134 

> If all works done, the issue mentioned above should be closed.

### Change Form
**Fill this up (NA if not available). If a certain criteria is not met, can you please give a reason.**

- [x] The pull request title has an issue number
- [x] The change works by "Smoke testing" or quick testing
- [x] The change has tests
- [ ] The change has documentation

## Other Information
**Small bug exists**
Do not merge this branch until all the bugs and progress were finished.
- Type issues
- Some Regex doesn't work, (e.g. valid username check with regex `^[\w.@+-]+$`)
- Some logic needs to be clarified (e.g. if photo was uploaded, progressbar goes or not?)
- Avataruploader needs further modification to fit penni design style (Optional ?) 
- I didn't use `<Form/>` because I notice that once I use it, some functions doesn't work, for example:
```tsx
//function...
  const handleEmailMobileChange = (e: ChangeEvent<HTMLTextTargetElement>) => {
    const value = e.target.value;
    setEmailMobile(value);
    setButtonVariant(value ? "default" : "inactive");
    if (isValidEmail(value) || isValidMobile(value)) {
      setErrorMessage(null);
      setButtonVariant("default");
    } else {
      setErrorMessage(errorMessagePreset);
      setButtonVariant("inactive");
    }
  }; // it doesn't work, I don't know why

return
       <Form onSubmit={()=>{example}}>
          <SingleLineInput
            name="emailMobile"
            onChange={handleEmailMobileChange}
            required={true}
            value={emailMobile}
            label="Email or mobile"
            type="text"
          /><Form/>
{/*middle part....*/}
        <Button
          className="h-[56px] w-full px-4 pb-4"
          variant={buttonInputVariant}
          type="submit"
          disabled={buttonInputVariant === "inactive"}
        >
          Continue
        </Button>
```

# Related issue

- Resolve #65

SeiKasahara committed:
store all works

SeiKasahara committed:
Merge remote-tracking branch 'origin' into issue-70-Request_to_Add_More_Button_Variant_and_Create_a_Button_Reference_Example_Page

SeiKasahara committed:
Merge remote-tracking branch 'origin' into issue-65-Replace_all_font_styling_with_global_font_classes

SeiKasahara committed:
reworked the main sign-up page

SeiKasahara committed:
finished basic functions to test

ErikaKK commented on client/src/components/ui/signup/main-page.tsx:
```diff
@@ -0,0 +1,134 @@
+          </div>
+        )}
+        <Button
+          className="h-[56px] w-full px-4 pb-4"
```
Pls remove padding. This makes the text not vertically center
<img width="292" alt="image" src="https://github.com/user-attachments/assets/2c75d3e8-dae5-4ced-a8e9-8aae9cee0406">

SeiKasahara replied:
Replaced all button to normal and proper style.

ErikaKK commented on client/src/components/ui/signup/main-page.tsx:
```diff
@@ -0,0 +1,134 @@
+        {errorMessage === errorMessagePreset ? (
+          <div className="callout text-red-600">{errorMessage}</div>
+        ) : (
+          <div className="callout text-red-600">
```
Pls use ErrorCallout instead

ErikaKK commented on client/src/components/ui/signup/avatar.tsx:
```diff
@@ -0,0 +1,178 @@
+            </button>
+          </div>
+
+          {showOptions && (
```
This showOptions should be set to false after upload img

ErikaKK commented on client/src/components/ui/signup/pw.tsx:
```diff
@@ -0,0 +1,102 @@
+    setRepeatPassword(value);
+    setRepeatPasswordVariant(e.target.value ? "default" : "inactive");
+    if (passwordValidation(password, value)) {
+      // 修改此行
```
<img width="560" alt="image" src="https://github.com/user-attachments/assets/7a15b6e8-1542-46ff-926c-da1088f9e8f8">
Should add password rule message and validation. Pls ask backend how they designed it

SeiKasahara replied:
They didn't have a clear criteria. I need to learn it by myself.

ErikaKK commented on client/src/pages/signup.tsx:
```diff
@@ -1,142 +1,135 @@
+          setTitle("Welcome!");
+          setClient("Poster");
+          console.log("Successfully created a new user");
+          window.location.href = "/index";
```
There is no `/index` page. You should direct them to `/poster` or `/bidder` based on their role. Pls use `router.push()` to redirect

ErikaKK commented on client/src/pages/signup.tsx:
```diff
@@ -1,142 +1,135 @@
-  ];
-  const multistepForm: MultistepForm = useMultistepForm(forms);
+      try {
+        const response = await api.post("/app/register/", data);
```
In the sign up page, you should send two post requests. One is sending to "/app/register/", another one is sending to "/app/profiles/". @yunho7687 Is this how it works?
<img width="1101" alt="image" src="https://github.com/user-attachments/assets/38da18b7-16aa-45e2-b6b0-2110badc7e81">

<img width="1325" alt="image" src="https://github.com/user-attachments/assets/26b629b8-c507-4674-ad89-07b59de57d2f">

SeiKasahara replied:
`/app/profiles` was destoryed. According to @yunho7687

ErikaKK reviewed (changes requested):
Pls remove padding style in all buttons. Pls see my comments. Pls check the reason for your type check failure

SodaVolcano commented on client/src/components/ui/signup/avatar.tsx:
```diff
@@ -0,0 +1,178 @@
+            <div className="fixed inset-0 z-50 flex items-end justify-center bg-black bg-opacity-50 transition-transform">
+              <div className="duration-800 animate-slide-up w-full space-y-1 rounded-t-lg p-4 ease-in-out">
+                <Button
+                  className="flex h-[56px] w-full px-4"
```
is that `[...]` I see???? don't use it pls :)

I'll ignore the rest to not clutter up the comments but pls change other occurrences as well

SodaVolcano commented on client/src/components/ui/signup/avatar.tsx:
```diff
@@ -0,0 +1,178 @@
+          )}
+        </div>
+      </div>
+      <div className="absolute left-0 right-0 top-[574px] flex flex-col gap-3 p-[1px_0] px-4">
```
same for this, not sure why you would need such a large value as `[574px]`, maybe a constant pixel value is not the right choice (i.e. look into using `vw` etc. For example, the sign in page used top padding that didn't use a specific pixel size. and what's `[1px_0]`???

SodaVolcano commented on client/src/components/ui/signup/check-unique.ts:
```diff
@@ -0,0 +1,37 @@
+      throw new Error(`Error in setting up request: ${axiosError.message}`);
+    }
+  }
+};
```
What is this function used for? If it's to check if user name already exist in the database, that's the backend's job. use the API `users/<user_id` to check if server responded with the user info (meaning it exist in database). Don't think the API exists so make an issue and assign me and I'll add it to the backend.

SodaVolcano replied:
Alternatively you can add it yourself, just need to edit `url.py` and `views.py`

SeiKasahara replied:
That makes sense I can made it by myself

SodaVolcano commented on client/src/components/ui/signup/main-page.tsx:
```diff
@@ -0,0 +1,134 @@
+      </div>
+    </div>
+  );
+};
```
For `<Form>`, I was stupid and had this line in all Input components inside `input.tsx`

```ts
onChange = context?.onChange ?? onChange ?? ((e) => {});
```

So, this is going to check if the `context` (the `<Form>` component) will supply the `onChange` method. If not, it falls back on the `onChange` in the prop. Since you wrapped it inside `<Form>` it'll always use the `onChange` provided by the form. To fix this, just change that line to

```js
onChange = onChange ?? context?.onChange ?? ((e) => {});
```

and you should be able to override the `onChange` function inside `<Form>`

SodaVolcano replied:
Also, consider moving validation for some inputs to the `onSubmit` function for the form. This is pretty much up to you whether you think it belongs in the `onChange` functions or in the `onSubmit` function, could make code more concise by letting `<Form>` manage the state (but the validation is only done when user clicks the button)

SodaVolcano replied:
UPDATE: do the same with `value` if the logic for that is the same (i.e. instead of checking for `context` first, it should check for `props.value` first

SodaVolcano commented on client/src/components/ui/signup/pw.tsx:
```diff
@@ -0,0 +1,102 @@
+    return false;
+  }
+  return true;
+}
```
Change to `return repeatedPassword == password`

SodaVolcano commented on client/src/components/ui/signup/pw.tsx:
```diff
@@ -0,0 +1,102 @@
+    setRepeatPassword(value);
+    setRepeatPasswordVariant(e.target.value ? "default" : "inactive");
+    if (passwordValidation(password, value)) {
+      // 修改此行
```
Remove Chinese comment, not everyone can read it :(

Edit: okay everyone working on penni right now probably can read it but avoid having such comments in future projects

yunho7687 replied:
haha bro just forgot to delete em lol

SodaVolcano commented on client/src/components/ui/signup/pw.tsx:
```diff
@@ -0,0 +1,102 @@
+      </div>
+    </div>
+  );
+};
```
You should be able to use `<Form>` here with the suggested change in above comment.

SodaVolcano commented on client/src/components/ui/signup/regex.ts:
```diff
@@ -0,0 +1,7 @@
+);
+export const mobileRegex = new RegExp("^[0-9]{10}$");
+
+export const userNameRegex = new RegExp("^[w.@+-]+$");
```
Why is there a file for just regex? Should be in the file that they are used in

SeiKasahara replied:
I'm thinking, in the future, if people want to validate more forms, they can extract regular expressions from this file. I actually want to move this to another location instead of keeping it separately in the `components/ui/signup` folder.

SodaVolcano replied:
Hm.. it doesn't make sense to separate this into its own file, regex strings are nothing more than just variables and we don't usually separate variables out into its own file (unless the variable is a global constant/configuration that affects the whole project)

SeiKasahara committed:
Merge remote-tracking branch 'origin' into issue-65-Replace_all_font_styling_with_global_font_classes

SeiKasahara committed:
On branch issue-65-Replace_all_font_styling_with_global_font_classes
 Changes to be committed:
	modified:   client/src/components/ui/callout.tsx
	modified:   client/src/components/ui/signup/avatar.tsx
	modified:   client/src/components/ui/signup/bio.tsx
	modified:   client/src/components/ui/signup/check-unique.ts
	modified:   client/src/components/ui/signup/main-page.tsx
	modified:   client/src/components/ui/signup/pw.tsx
	modified:   client/src/components/ui/signup/regex.ts
	modified:   client/src/components/ui/signup/welcome.tsx
	modified:   client/src/pages/signup.tsx
	modified:   server/app/urls.py
	modified:   server/app/views.py

SeiKasahara commented:
# Enhancement Log

1. Designed a pop-up animation for the error callout on the signup page without affecting the original layout.
2. The issues with types and regular expressions have been resolved.
3. Using `app/register` for email uniqueness validation caused users to be registered directly during the validation phase, resulting in a 400 status code on subsequent registration attempts. Therefore, a custom validate API was created to allow a separate GET API for email uniqueness validation, with URL injection prevention. A separate POST API is used to prevent password brute-force attacks (by using the common password validation).
4. Added a token to local storage for users to directly access the corresponding page after registration.
5. The avatar can now be captured using the camera.
6. Modified most of the TailwindCSS styles, using `h-number` instead of `h-[number px]`.
7. Added back and skip buttons on the progress bar for users to set content in the future.
8. Basically all anomaly incident will be prevent in the frontend part. they must pass the frontend validation to access the final submission.

ErikaKK commented on client/src/components/ui/signup/pw.tsx:
```diff
@@ -0,0 +1,177 @@
+    inputCheck(value, setPasswordVariant);
+  };
+
+  const handleRepeatPasswordChange = (
```
This one doesn't check when changing the first password input field.
<img width="1303" alt="image" src="https://github.com/user-attachments/assets/c0df3ddf-534b-474b-875a-9bfaf5c82410">
Needs to add to the first pw input field

ErikaKK commented on client/src/components/ui/signup/pw.tsx:
```diff
@@ -0,0 +1,177 @@
+    password: string,
+    hook: React.Dispatch<React.SetStateAction<string>>,
+  ) {
+    if (password.length <= 8) {
```
change to `password.length < 8`

ErikaKK commented on client/src/components/ui/signup/pw.tsx:
```diff
@@ -0,0 +1,177 @@
+    if (password.length <= 8) {
+      setErrorMessage(
+        "This password is too short. It must contain at least 8 characters.",
+      );
```
Password rules should be more detailed. e.g. not too common

SeiKasahara replied:
The common validation part will be validated in backend

ErikaKK commented on client/src/pages/signup.tsx:
```diff
@@ -1,142 +1,168 @@
+        email: emailMobile,
+        username: fullName,
+        password: password,
+        //avatar: profilePhoto,
```
change to `avatar_url` and uncomment this

ErikaKK commented on client/src/components/ui/signup/main-page.tsx:
```diff
@@ -0,0 +1,125 @@
+const uniqueErrorMessage = (
+  <>
+    Email or Mobile has been registered, please{" "}
+    <Link className="text-penni-main" href="/sign-in">
```
Pls change `/sign-in` to `/login` bc I changed the name of sign-in page

ErikaKK reviewed (changes requested):
Just needs a bit change.  Pls see my comments

SodaVolcano commented on client/src/components/ui/callout.tsx:
```diff
@@ -1,7 +1,9 @@
 
 interface CalloutProps {
-  text: string;
+  text: React.ReactNode;
```
?????? why is this being changed???

SodaVolcano replied:
If you want links in the callout, set `onClick`, this will probably break where ever `Callout` is used

SeiKasahara replied:
errorMessage text can be null, can I change to `string | null` ?

SodaVolcano replied:
no it shouldn't be nullable, was it modified somewhere?

SeiKasahara replied:
I notice that if it’s empty string the error-callout component still takes position, but if the text is null it won’t.Since user input is correct, the error-callout should be hided. So I tried to set error messages as <string | null>

SodaVolcano replied:
If the error callout should be hidden then you shouldn't be rendering it at all. Don't hide it by messing with the string parameter, hide it by not rendering it. Other pages kept a state `errorOccured` and if it's set to `true` then the error callout is rendered

SodaVolcano commented on client/src/components/ui/signup/avatar.tsx:
```diff
@@ -97,27 +71,86 @@ export const Ava: React.FC<AvaProps> = ({
-        {/*Avatar Uploader*/}
-        <div className="flex items-center justify-center">
+        <div className="mt-2 flex">
           <div className="relative h-[136px] w-[136px]">
```
`h-[136px] w-[136px]` change

SodaVolcano commented on client/src/components/ui/signup/avatar.tsx:
```diff
@@ -97,27 +71,86 @@ export const Ava: React.FC<AvaProps> = ({
+              <div className="fixed inset-0 z-50 flex flex-col items-center justify-center bg-black bg-opacity-70 transition-transform">
+                <div className="mt-56 flex flex-col items-center justify-center">
+                  <Webcam
+                    className="h-[136px] w-[136px] rounded-full border-2 object-cover"
```
`h-[136px] w-[136px]`

SodaVolcano commented on client/src/components/ui/signup/check-unique.ts:
```diff
@@ -6,18 +6,19 @@ type AxiosCustomError = AxiosError<{
-      if (profile.full_name === input) {
-        return true;
-      }
+    // console.log(data);
```
remove if you are commenting this out

SodaVolcano reviewed (changes requested):
Yeah nah I'm not gonna review the rest, this project is near the end so I'm not gonna check the code that thoroughly. If you don't want to change the code you can dismiss this change request on GitHub

ErikaKK commented:
FYI, This is the new fields (plus `avatar_url`) that you have to submit when adding a user
<img width="693" alt="image" src="https://github.com/user-attachments/assets/7bfedf02-2f47-4d76-818c-c5d824956a41">

SeiKasahara commented:
> FYI, This is the new fields (plus `avatar_url`) that you have to submit when adding a user <img alt="image" width="693" src="https://private-user-images.githubusercontent.com/98962967/353526286-7bfedf02-2f47-4d76-818c-c5d824956a41.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjIzNTQ2NjYsIm5iZiI6MTcyMjM1NDM2NiwicGF0aCI6Ii85ODk2Mjk2Ny8zNTM1MjYyODYtN2JmZWRmMDItMmY0Ny00ZDc2LTgxOGMtYzVkODI0OTU2YTQxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MzAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzMwVDE1NDYwNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWJhMDFjYTRlMTRkOWY5MGEyZTY2YWU5MjRkYTc4ZjI2N2UwYWY2ZmQ5MjAzY2FkNGRkMzJmZGMxNzM5OTQ5NDkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.8hiPLRJbHFQqYaHBzFHDQj7MzbrZXbs1rJN5-a43RVc">

is that avatar data requirement new added?

SeiKasahara committed:
finished all required field

yunho7687 committed:
Merge remote-tracking branch 'origin' into issue-65-Replace_all_font_styling_with_global_font_classes

yunho7687 committed:
chore: Refactor code to remove unnecessary lines and improve readability

yunho7687 commented:
@SodaVolcano thanks for all your dedication! 🐐 you are the real one!

yunho7687 commented:
@ErikaKK @SeiKasahara thanks for your dedication!

yunho7687 committed:
Refactor Callout component to require non-null text prop

SodaVolcano commented:
> @SodaVolcano thanks for all your dedication! 🐐 you are the real one!

@ErikaKK should get the credit since she helped out waaaaay more than me HAHA  🫡🫡🫡🫡🫡🎉🎉🎉 absolute commitment!!

SeiKasahara commented:
> @ErikaKK @SeiKasahara thanks for your dedication!

Great Gig too!

yunho7687 committed:
Refactor signup components to improve code readability and error handling

yunho7687 commented:
several convention issues but run out of time :(
